### ISSUE TITLE
Make unconfigured methods that return type Lazy<T> and Task<T> behave more like methods that return T

### ISSUE THREAD

**Issue by HenrikPoulsen** [FakeItEasy/FakeItEasy#560]
Version: 1.25.3
So I noticed today when I tried this framework for the first time that it hangs when the code calls a method with the signature:
`Task<QueueDescription> CreateQueueAsync(string queueName);`
Checking the output I can see:
```
Exception thrown: 'System.MissingMethodException' in mscorlib.dll
Exception thrown: 'System.MissingMethodException' in mscorlib.dll
Exception thrown: 'Castle.DynamicProxy.InvalidProxyConstructorArgumentsException' in FakeItEasy.dll
Exception thrown: 'System.MissingMethodException' in mscorlib.dll
Exception thrown: 'System.MissingMethodException' in mscorlib.dll
Exception thrown: 'Castle.DynamicProxy.InvalidProxyConstructorArgumentsException' in FakeItEasy.dll
Exception thrown: 'System.ArgumentException' in mscorlib.dll
Exception thrown: 'Castle.DynamicProxy.ProxyGenerationException' in FakeItEasy.dll
Exception thrown: 'Castle.DynamicProxy.Generators.GeneratorException' in FakeItEasy.dll
Exception thrown: 'System.ArgumentException' in FakeItEasy.dll
Exception thrown: 'Castle.DynamicProxy.Generators.GeneratorException' in FakeItEasy.dll
Exception thrown: 'Castle.DynamicProxy.Generators.GeneratorException' in FakeItEasy.dll
Exception thrown: 'System.ArgumentException' in FakeItEasy.dll
Exception thrown: 'System.ArgumentException' in FakeItEasy.dll
```

After wasting some time on it the problem seems to be because QueueDescription has an internal constructor, and for some reason FakeItEasy is trying to instantiate an instance rather than just returning null.
Here is a simple way of reproducing the issue:
```c#
[TestMethod]
public async Task Test_BreakIt()
{
    var breaker = A.Fake<IFakeItEasyBreaker>();
    await breaker.BreakIt();
}

public interface IFakeItEasyBreaker
{
    Task<ClassWithInternalConstructor> BreakIt();
}

public class ClassWithInternalConstructor
{
    internal ClassWithInternalConstructor()
    {
    }
}
```

In the steps to reproduce I could of course add:
`[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]`
But it does not help with my actual issue, since QueueDescription is a part of Microsoft.ServiceBus.dll which I don't have much power over.
Then I can of course setup so that i tell it to explicitly return null, but coming from Moq this has never been an issue. The test I got this problem with was a previous Moq test which worked fine with Moq, it returned null as expected.

**Comment by blairconrad** [comment]
Hi, @HenrikPoulsen. Thanks for writing.

I'm a little surprised that the call hangs (but then again I don't use `await` very often, so we may just be hitting a hole in my knowledge, something that happens all too frequently). I will take a look.
In the meantime, I _can_ shed some light on one comment you made

> for some reason FakeItEasy is trying to instantiate an instance rather than just returning null

This is by design. FakeItEasy's default behaviour for `Task<T>` return types is to return a completed task whose result is a Dummy `T`. This is (except for the "Dummy" part, which I should amend) explained at https://github.com/FakeItEasy/FakeItEasy/wiki/Default-fake-behavior

As you point out, you can configure the method to return `null` instead, which will work around the problem in the meantime.

**Comment by blairconrad** [comment]
Since `QueueDescription` does not have an accessible constructor, FakeItEasy can't make a Dummy of it. So it falls back on other, non-Task-specific, strategies to make a Dummy `Task<QueueDescription>`. These involve trying to create a Fake `Task<QueueDescription>`, which ultimately succeeds, but this requires FakeItEasy to try to find an acceptable `Task<QueueDescription>` constructor, [of which there are many](https://msdn.microsoft.com/en-us/library/dd321477(v=vs.110).aspx), and some attempt to use the constructors fail. That's why you see the exceptions you did.

Ultimately, a Fake `Task<QueueDescription>` is created, but due to it being unconfigured, it returns default values for all methods and properties, so it looks like an uncompleted Task. I believe this is why you see the hang—the test is effectively waiting on a task that will never complete.

I hope that clears some things up.

Ultimately, if the default behaviour of faked `IFakeItEasyBreaker` does not work for the test, I think the answer is to configure the `IFakeItEasyBreaker`.

**Comment by HenrikPoulsen** [comment]
Assuming this works:
`ClassWithInternalConstructor BreakIt();`
Without the Task, then I would say the expected behaviour would be that it should work with the Task as well.
In a world where every unit test is perfect and all code is written according to S.O.L.I.D there would be no issue with manually faking every single call you want faked in the method you are testing, since there would be so few.
In the real world that is seldom so, you often find tests which tests way too much and have too many dependencies with lots of small methods peppered everywhere which are inconsequential for the test at hand.

However in this scenario FakeItEasy is doing unexpected things. One of the calls in the following snippet hangs, and the other does not, and it is in no way apparent at a glance why they should be treated differently.
```
if (!await namespaceManager.QueueExistsAsync(queueName))
{
     await namespaceManager.CreateQueueAsync(queueName);
}
```
And I assume it would work flawlessly without the Tasks, so it has very inconsistent behavior, which is always undesirable in any framework.

**Comment by blairconrad** [comment]
Hi, @HenrikPoulsen. In no particular order:

> In a world where every unit test is perfect and all code is written according to S.O.L.I.D there would be no issue with manually faking every single call you want faked in the method you are testing, since there would be so few.
> In the real world that is seldom so, you often find tests which tests way too much and have too many dependencies with lots of small methods peppered everywhere which are inconsequential for the test at hand.

I'm not really sure how this influences the matter at hand. Are you saying that your codebase has many calls to methods that return `Task<ClassWithInternalConstructor>` and faking the return values in each case is inconvenient? I sympathize with your pain. While we look for a resolution for this issue, you could ease your pain by using an [IDummyDefinition](https://github.com/FakeItEasy/FakeItEasy/wiki/Custom-Dummy-Creation-in-version-1.x#how-it-works) to return an acceptable default value for all calls:

```c#
public TaskOfClassWithInternalConstructorDummyDefinition: IDummyDefinition
{
    Type ForType { get { return typeof(Task<ClassWithInternalConstructor>); }
    object CreateDummy() { return /* your preferred Dummy goes here */ };
}
```

> Assuming this works:
> `ClassWithInternalConstructor BreakIt();`
> Without the Task, then I would say the expected behaviour would be that it should work with the Task as well.

When faced with `ClassWithInternalConstructor BreakIt()`, FakeItEasy would try as hard as it can to return a non-null Dummy `ClassWithInternalConstructor`. Ultimately, it would be unable to, and would return `null`. Whether that's considered "working" is perhaps a matter for interpretation. Nevertheless, it's at least established behaviour that's been in FakeItEasy for a very long time, and I think it's reasonable to use it as a model.

> Without the Task, then I would say the expected behaviour would be that it should work with the Task as well.

One way to interpret that sentence would be to say that `Task<ClassWithInternalConstructor> BreakIt();` works as well. In this case, FakeItEasy tries has hard as it can to return a non-null Dummy `Task<ClassWithInternalConstructor>`. It does this by trying to construct a concrete completed Task whose result is a non-null Dummy `ClassWithInternalConstructor`. Just as before, it's unable to create the `ClassWithInternalConstructor`, so it falls back to other methods for creating the Dummy, including seeing if it can make a `Lazy` to satisfy the request (it can't), and then trying to make a Fake `Task<ClassWithInternalConstructor>`, which ultimately succeeds, so we get the unconfigured Fake `Task<ClassWithInternalConstructor>`, which happens to look like an uncompleted Task.
Is this desirable? Perhaps not. Then again, getting a null `ClassWithInternalConstructor` may not have been desirable in the first case. I'd say neither are ideal.

So, whether the behaviour is consistent depends a little on your frame of mind. In each case, FakeItEasy is following the [Rules of how to make a Dummy](https://github.com/FakeItEasy/FakeItEasy/wiki/Dummies#how-are-the-dummies-made) (note that I updated the documentation _just now_—before the Task and Lazy notes neglected to specify that those rules only apply when the "target" type of the Task and Lazy could be made into a Dummy).
I'd say that if the reader could glean that the `ClassWithInternalConstructor` cannot be made into a Dummy, she would understand what's going on. Does that make it right? Probably not.

I do like the idea that an unconfigured `Task<TResult>`-returning method could return a completed `Task<TResult>` where the `Result` property is the same value that would be got from a `TResult`-returning method. (And similarly for `Lazy<TValue>`.)

I am not 100% sure that this is what you originally expected: 

> Then I can of course setup so that i tell it to explicitly return null, but coming from Moq this has never been an issue. The test I got this problem with was a previous Moq test which worked fine with Moq, it returned null as expected.

Sounds to me like you expected a null `Task<ClassWithInternalConstructor>`, but I am instead proposing that the faked `BreakIt` would return a completed `Task<ClassWithInternalConstructor>`, where the `Result` property's value was `null`. I think this is more consistent behaviour overall, and the preferable result to returning `null`. So the Dummies documentation would probably read a little more like

---- 
When FakeItEasy needs to access a Dummy of type `T`, it tries a number of approaches in turn, until one succeeds:

1. see if there's a user-supplied [[custom Dummy creation]] mechanism for `T`
1. if `T` is `Task`, the returned Dummy will be an actual `Task` that completes immediately<sup>1</sup>
1. if `T` is `Task<TResult>`, the returned Dummy will be an actual `Task<TResult>` that completes immediately<sup>1</sup> and whose `Result` is the same as that returned by an unconfigured method with return type `TResult`.
1. if `T` is a `Lazy<TValue>`, the returned Dummy will be an actual `Lazy<TValue>` whose `Value` is the same as that returned by an unconfigured method with return type `TValue`<sup>2</sup>
1. if `T` is [fakeable](https://github.com/FakeItEasy/FakeItEasy/wiki/What-can-be-faked), the Dummy will be a Fake `T`
1. if `T` is a value type, the Dummy will be a `T` created via `Activator.CreateInstance`
1. if nothing above matched, then `T` is a class. Loop over all its constructors in _descending order of argument list length_.  
For each constructor, attempt to get Dummies to satisfy the argument list. If the Dummies can be found, use `Activator.CreateInstance` to create the Dummy, supplying the Dummies argument list. If the argument list can't be satisfied, then try the next constructor.

If none of these strategies yield a viable Dummy, then FakeItEasy can't make a Dummy of type `T`.

----

If you're agreed, I think we can mark this issue as an enhancement, put it in the backlog for FakeItEasy 2.0.0 and give the other owners a chance to weigh in.

**Comment by HenrikPoulsen** [comment]
I guess I was a bit vague, and misunderstood some things in your initial responses.
When I mentioned I expected it to return null, I should of course have indicated that I meant a Task.FromResult<ClassWithInternalConstructor>(null); when for some reason it cannot instantiate a dummy for it.

Also when you said
> Ultimately, if the default behavior of faked IFakeItEasyBreaker does not work for the test, I think the answer is to configure the IFakeItEasyBreaker.

I Interpreted that as "By design. Wont fix", which is what most of my response was trying to explain why this should be considered a defect and not intended behavior.

The reasoning is if the following works (returns null/dummy and doesn't hang or throw exceptions or whatever the case may be)
```breaker.BreakIt() // Signature return type ClassWithInternalConstructor```
and this doesn't (in this case it hangs permanently, not even an exception)
```await breaker.BreakIt(); // Signature return type Task<ClassWithInternalConstructor>```

That feels a bit inconsistent. And coming from Moq that's just something you've grown to expect. I would expect you see this in other frameworks as well.
Things will just move along as long as I am not trying to use the return value of my await breaker.BreakIt() since it will most likely be null (which it would be in Moq).

The code example I showed in my previous reply works with Moq. It returns A completed task with null in it, so it's not like I'm trying to convince you to start doing Voodoo :)

tl;dr
It's fine with me if you mark it as an enhancement.

**Comment by blairconrad** [comment]
Sir.

> > Ultimately, if the default behavior of faked IFakeItEasyBreaker does not work for the test, I think the answer is to configure the IFakeItEasyBreaker.

> I Interpreted that as "By design. Wont fix", which is what most of my response was trying to explain why this should be considered a defect and not intended behavior.

A reasonable interpretation, as that was the direction I was leaning. I was rushing a little when I typed that out (never a good idea), and probably should've reworded. And also been clear that even when I think that things should work a certain way, I'm always open to being convinced. Your response did a much better job of that than the initial post.

And one of my big goals of my initial response was to point out what was actually going on - the title "Task<> return values with internal constructors hangs" suggests that FakeItEasy hangs while creating a Task rather than that the Task is created as not completed. I should've realized that you knew the Task was returned, but that awaiting hangs.

> it's not like I'm trying to convince you to start doing Voodoo
Indeed not.

**Comment by blairconrad** [comment]
@FakeItEasy/owners, I think @HenrikPoulsen has identified a good improvement for our built-in `Lazy` and `Task` Dummy creation. Any objections to the proposed fix?

@HenrikPoulsen, if there are no objections to the proposed improvement, is this something you're interested in working on?

**Comment by adamralph** [comment]
Do I understand correctly:

* When a fake method returns a reference type `T` which has an internal constructor only, it returns `null`
* When a fake method returns `Task<T>` (with `T` as above), it blows up with these exceptions?

**Comment by blairconrad** [comment]
1. Yes.
2. Almost? When we return the Dummy `Task<T>`, it's actually an unconfigured Fake, which presents as a not-completed Task (because IsCompleted returns false) so the `await` hangs. All the exceptions listed occur because FakeItEasy is trying out base constructors to make a fake `Task`. But FakeItEasy catches each exception, rejects that constructor and just moves on to the next, before ultimately finding one that works.

**Comment by blairconrad** [comment]
… so @HenrikPoulsen is proposing that the Dummy Task instead be completed, and created with the same Result we would've gotten had we called an analogous non-Tasky method.

**Comment by adamralph** [comment]
Oh dear...

I agree with the proposal.

**Comment by blairconrad** [comment]
Marking as up-for-grabs, as altering the behaviour of `ResolveByCreatingTaskStrategy` and `ResolveByCreatingLazyStrategy` to return a `Type.GetDefaultValue`
when `TryResolveDummyValue` fails shouldn't be too onerous.

**Comment by blairconrad** [comment]
Nice work, @thomaslevesque.

**Comment by blairconrad** [comment]
Thanks for raising this, @HenrikPoulsen. Look for your name in the release notes!

:trophy:

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
